### PR TITLE
Add KPool DSA indexer, NoPE MLA support, and FP8 precision controls for the indexer and MLA attention

### PR DESCRIPTION
**Pull request by HollowMan6**
- [x] I, the PR author, have personally reviewed every line of this PR.

# What does this PR do?
Add the KPool (pooled-key) DSA indexer used by [GLM-5.3-Flash](https://huggingface.co/zai-org/GLM-5.3-Flash), NoPE support for the DSA indexer and absorbed MLA, and the precision controls that keep the indexer and the MLA attention GEMMs unquantized under FP8 training.

Split out of #7054 (GLM-5.3-Flash support) as discussed there; #7054 now carries only the KDA layers. Sibling PRs: #7521 (mHC precision controls) and #7523 (blockwise FP8 wgrad guard). All are independent of each other and based on `main`.

GLM-5.3-Flash's indexer uses `index_kpool = 4`, `index_kpool_always_select_tail = true` and `index_topk = 2048` over NoPE MLA (`qk_rope_head_dim = 0`).

## Changes

- `dsa_indexer_kpool > 1`: keys are compressed `kpool` tokens at a time with a per-token compression gate (`index_kpool_compress_gate`, BF16, same parameter name as the checkpoint) plus a per-slot additive position bias (`index_kpool_compress_ape`, FP32): per-dimension softmax over the slots, FP32 accumulation, BF16 pooled key. Top-k runs over pools (`index_topk // kpool`) and is expanded back to token indices (`fused_qk_topk_kpool`); `dsa_indexer_kpool_always_select_tail` appends each query's incomplete causal pool so short prefixes stay fully visible. Packed THD restarts pools at document boundaries. The fused cuDNN DSA path is bypassed for KPool, and the head-weight projection stays in FP32 (cast to the query dtype where the cuDNN BF16 indexer kernel is used).
- `dsa_indexer_kpool_fp8`: emulate the serving indexer's FP8 input path (FP32 Hadamard → BF16 → E4M3 with a power-of-two row scale) on `q` and on the pooled keys, so training top-k matches inference.
- NoPE: `qk_pos_emb_head_dim == 0` is supported by `DSAIndexer` and `AbsorbedMLA` (no rotary embedding is built; the absorbed-q path has no positional slice).
- DSA indexer precision split under FP8: `linear_wk` and `linear_weights_proj` are built under `get_fp8_disabled_context(config, is_init=True)` (BF16 parameters, no FP8 scales), and in `forward_before_topk` the k path (`linear_wk`, `k_norm`, RoPE, activation rotation), `linear_weights_proj`, the `linear_wq_b` GEMM and the KPool compression gate run under `get_fp8_disabled_context`. Net effect: the whole indexer runs unquantized while the rest of the model keeps FP8.
- `mla_disable_attention_fp8`: run `linear_q_down_proj`, `linear_q_up_proj`, `linear_kv_down_proj` and `linear_proj` of `AbsorbedMLA` with FP8 disabled.

## Motivation

With an FP8 actor and a vLLM rollout, every module the rollout computes in BF16 has to be computed in BF16 on the actor as well, otherwise the two diverge. The released FP8 checkpoint keeps the whole DSA indexer (`wq_b`, `wk`, `weights_proj`, `k_norm`, `index_kpool_compress_*`) in BF16 (`quantization_config.modules_to_not_convert`), and vLLM runs the main MLA attention path in BF16. The indexer matters most: it feeds a discrete top-k, so FP8 noise there changes *which* tokens are attended, and actor/rollout agreement (Pearson) degrades. These options let the actor reproduce the serving precision split while the MoE experts (the bulk of the FLOPs) stay in FP8.

## Relationship to other PRs

- Related: #6842 (main: DSA indexer `weights_proj` precision controls), #6877 (dev: MLA NoPE), #6392 (GLM-5.2 training support tracking).

## Testing

- `tests/unit_tests/transformer/experimental_attention_variant/test_kpool_causal_tail.py` (new): every query keeps its full causal prefix under the pool budget (packed / unpacked, strided queries, FP8 on / off); the FP8 input emulation matches a Hadamard-matrix reference.
- `tests/unit_tests/transformer/experimental_attention_variant/test_attention_variant_dsa.py`: KPool projection dtypes and backward.

```bash
pytest tests/unit_tests/transformer/experimental_attention_variant/test_kpool_causal_tail.py \
       tests/unit_tests/transformer/experimental_attention_variant/test_attention_variant_dsa.py
```

:warning: For major changes (either in lines of code or in its impact), please make sure to first share a design doc with the team. If you're unsure what's the best way to do so, contact @NVIDIA/mcore-oncall.

## Issue tracking

For PRs from open-source community contributors:

- **New features**: a linked issue is **required**. Please open a [feature request](https://github.com/NVIDIA/Megatron-LM/issues/new?template=feature_request.md) and reference it here before submitting the PR.
- **Small updates (bug fixes, minor improvements)**: a linked issue is **recommended** and will accelerate the PR review process.

Linked issue: Related to #6392

### Pre-checks

- [x] I have added relevant unit tests
- [ ] I have added relevant functional tests
- [ ] If this PR adds or changes a GPU kernel (Triton, `jit_fuser`/`torch.compile`, CUDA extension, TE or external-library dispatch, or a scatter/index accumulation), I have added or updated its bit-exact determinism test and registered it in `tests/unit_tests/determinism/kernels/manifest.py` ([guide](https://github.com/NVIDIA/Megatron-LM/blob/main/docs/developer/determinism/testing.md))
- [x] I have added proper typing to my code [Typing guidelines](https://docs.python.org/3/library/typing.html)
- [ ] I have added relevant documentation
- [x] I have run the [autoformatter.sh](https://github.com/NVIDIA/Megatron-LM/blob/main/tools/autoformat.sh) on my PR

### Code review

Feel free to message or comment @NVIDIA/mcore-oncall to help accelerate your merge into main. The less complex your PR is, the faster it will be approved and merged!

All PRs start as **draft**. If you open a non-draft PR, it will be automatically converted to draft.

#### Step 1: Mark PR as "Ready for Review"

1. When your PR is ready, click **Ready for Review**.
2. An oncall reviewer is auto-assigned and expert reviewers are notified based on your changes.
   - Some PRs may jump straight to step 2. This is determined by `.github/CODEOWNERS`.

:warning: Only mark as ready once merge-conflicts are resolved and the CI is passing.
Final Review might get declined if these requirements are not fulfilled.

#### Step 2: Final Review

For PRs that change `megatron/core`, once all expert reviewers have approved, the `Final Review` label is applied **automatically** and final reviewers are assigned.

For PRs outside `megatron/core`, this step is skipped.

#### Step 3: Approved

Once all required reviewers have approved, the `Approved` label is applied **automatically**.

### Merge

Any member of [mcore-engineers](https://github.com/orgs/NVIDIA/teams/mcore-engineers) will be able to merge your PR.

### CODE REVIEW

File: megatron/core/transformer/transformer_config.py [hunk 1]
```diff
@@ -379,12 +379,28 @@ class TransformerConfig(ModelParallelConfig):
...
+    """Match FP8 KPool index scores
+       model weights and attention remain in their configured dtype."""
+
     dsa_indexer_k_norm_epsilon: Optional[float] = None
     """Optional epsilon override for the DSA indexer key LayerNorm."""
 
     dsa_indexer_k_norm_fp32: bool = False
     """Whether DSA indexer key LayerNorm should run on fp32 inputs."""
 
+    mla_disable_attention_fp8: bool = False
```
Comment (buptzyb) [inline megatron/core/transformer/transformer_config.py]: This config should belong to `MLATransformerConfig`. Also, "disable fp8" is ambiguous. We don't know you mean to keep bf16 or fp4. Consider something like "mla_proj_disable_quantization" / "mla_proj_keep_in_bf16" / ...

File: megatron/core/transformer/transformer_config.py [hunk 2]
```diff
@@ -379,12 +379,28 @@ class TransformerConfig(ModelParallelConfig):
     dsa_indexer_scoring_relu: bool = True
     """Whether DSA indexer should apply ReLU to q@k^T scores before weighting."""
 
+    dsa_indexer_kpool_fp8: bool = False
```
Comment (buptzyb) [inline megatron/core/transformer/transformer_config.py]: Why don't we put this after `dsa_indexer_kpool`? Also, as commented in `mla_disable_attention_fp8`, try to use "quantization" instead of "fp8" in naming.

File: megatron/core/transformer/experimental_attention_variant/absorbed_mla.py [hunk 3]
```diff
@@ -468,7 +480,11 @@ def get_query_key_value_tensors(
             #     q_compressed: [s, b, q_lora_rank / TP]
             # elif linear_q_down_proj is Linear:
             #     q_compressed: [s / TP, b, q_lora_rank]
-            q_compressed, _ = self.linear_q_down_proj(hidden_states)
+            if self.config.mla_disable_attention_fp8:
```
Comment (buptzyb) [inline megatron/core/transformer/experimental_attention_variant/absorbed_mla.py]: If you want to avoid so many if-else, you can write `fp8_disabled_context = get_fp8_disabled_context(self.config) if self.config.mla_disable_attention_fp8 else nullcontext()` at the beginning, then just `with fp8_disabled_context:` everywhere.

File: megatron/core/transformer/experimental_attention_variant/dsa.py [hunk 4]
```diff
@@ -2318,7 +2684,29 @@ def slice_topk_to_local_sequence_parallel_rows():
             # ===================================
             # Get top-k indices
             # ===================================
-            if fused_bounds is not None:
+            if _is_kpool:
```
Comment (buptzyb) [inline megatron/core/transformer/experimental_attention_variant/dsa.py:2691]: kpool is only used without indexer loss? Is this how it's designed or it's just you didn't add the support yet?

File: megatron/core/transformer/experimental_attention_variant/dsa.py [hunk 5]
```diff
@@ -2318,7 +2684,29 @@ def slice_topk_to_local_sequence_parallel_rows():
...
+                        varlen_starts=varlen_starts,
+                        varlen_ends=varlen_ends,
+                        key_positions=key_positions,
+                        cu_seqlens_kv=cu_seqlens_kv if packed_thd else None,
+                        use_relu=self.config.dsa_indexer_scoring_relu,
+                        always_select_tail=self.indexer.index_kpool_always_select_tail,
+                        fp8_indexer=self.config.dsa_indexer_kpool_fp8,
+                    )
+                    del _index_scores
+                slice_topk_to_local_sequence_parallel_rows()
```
Comment (buptzyb) [inline megatron/core/transformer/experimental_attention_variant/dsa.py]: Why slice_topk_to_local_sequence_parallel_rows twice?

File: megatron/core/transformer/experimental_attention_variant/dsa.py [hunk 6]
```diff
@@ -1315,17 +1595,19 @@ def __init__(
...
-            self.index_head_dim,
-            config=self.config,
-            init_method=self.config.init_method,
-            bias=False,
-            skip_bias_add=False,
-            skip_weight_param_allocation=False,
-            parallel_mode="duplicated",
-        )
+        # wk + weights_proj run in BF16; they feed index scores directly.
+        with get_fp8_disabled_context(self.config, is_init=True):
```
Comment (buptzyb) [inline megatron/core/transformer/experimental_attention_variant/dsa.py]: No fp8 for all models?

File: megatron/core/transformer/transformer_config.py [hunk 7]
```diff
@@ -379,12 +379,28 @@ class TransformerConfig(ModelParallelConfig):
...
+    mla_disable_attention_fp8: bool = False
+    """Force MLA attention GEMMs (q_a_proj, q_b_proj, kv_a_proj, o_proj) to BF16
+    even under FP8 training. This aligns the actor's attention path with vLLM
+    rollout, which runs the main MLA attention in BF16. Only the attention GEMMs
+    (~2.5% of model params) are affected; MoE experts stay FP8."""
+
+    dsa_indexer_kpool: int = 1
+    """Number of keys per softmax-weighted indexer pool; 1 keeps per-token selection."""
+
+    dsa_indexer_kpool_always_select_tail: bool = False
```
Comment (buptzyb) [inline megatron/core/transformer/transformer_config.py]: Why not default to True?

File: megatron/core/transformer/experimental_attention_variant/dsa.py [hunk 8]
```diff
@@ -1338,17 +1620,39 @@ def __init__(
...
+            )
+
+        # Pool compression uses replicated per-token gates and slot-position biases.
+        self.index_kpool = int(self.config.dsa_indexer_kpool)
+        self.index_kpool_always_select_tail = bool(self.config.dsa_indexer_kpool_always_select_tail)
+        # Per-token gate score for the kpool path; set in forward_before_topk.
+        self._kpool_gate_score: Optional[torch.Tensor] = None
+        if self.index_kpool > 1:
+            # fp32 [kpool, index_head_dim] additive positional bias per pool slot.
+            self.index_kpool_compress_ape = torch.nn.Parameter(
```
Comment (buptzyb) [inline megatron/core/transformer/experimental_attention_variant/dsa.py:1648]: If you only use ape and gate inside `torch.no_grad`, you should define them as `requires_grad=False`.

File: megatron/core/transformer/experimental_attention_variant/dsa.py [hunk 9]
```diff
@@ -1430,42 +1742,63 @@ def forward_before_topk(
         # q linear and apply rope to q
         # =========================================
         # [seqlen, batch, q_lora_rank] -> [seqlen, batch, index_n_heads * index_head_dim]
-        q, _ = self.linear_wq_b(qr)
+        with get_fp8_disabled_context(self.config):
+            q, _ = self.linear_wq_b(qr)
```
Comment (buptzyb) [inline megatron/core/transformer/experimental_attention_variant/dsa.py:1752]: No fp8 indexer for all models? But you initialize it in fp8, why?

File: megatron/core/transformer/experimental_attention_variant/absorbed_mla.py [hunk 10]
```diff
@@ -489,7 +505,11 @@ def get_query_key_value_tensors(
         #     kv_combined: [s, b, (kv_lora_rank + qk_pos_emb_head_dim) / TP]
         # elif linear_kv_down_proj is Linear:
         #     kv_combined: [s / TP, b, (kv_lora_rank + qk_pos_emb_head_dim)]
-        kv_combined, _ = self.linear_kv_down_proj(hidden_states)
+        if self.config.mla_disable_attention_fp8:
+            with get_fp8_disabled_context(self.config):
+                kv_combined, _ = self.linear_kv_down_proj(hidden_states)
```
Comment (buptzyb) [inline megatron/core/transformer/experimental_attention_variant/absorbed_mla.py]: But why still initializing these proj in fp8?

File: megatron/core/transformer/experimental_attention_variant/dsa.py [hunk 11]
```diff
@@ -1338,17 +1620,39 @@ def __init__(
...
-            self.index_n_heads,
-            config=self.config,
-            init_method=self.config.init_method,
-            bias=False,
-            skip_bias_add=False,
-            skip_weight_param_allocation=False,
-            parallel_mode="duplicated",
-        )
+        with get_fp8_disabled_context(self.config, is_init=True):
+            self.linear_weights_proj = build_module(
```
Comment (buptzyb) [inline megatron/core/transformer/experimental_attention_variant/dsa.py]: You may want to build your PR upon #6842 because it allows you to choose the calculation and output precision for `linear_weights_proj`.

**Review by buptzyb — COMMENTED** [review]
I think we have to make sure our changes don't break existing models' behaviors. Thanks!

File: megatron/core/transformer/transformer_config.py (hunk 1 above)
Comment (HollowMan6) [reply megatron/core/transformer/transformer_config.py]: Resolved. The option now lives in MLATransformerConfig and is named mla_proj_disable_quantization, which makes the precision contract explicit. Its default remains False, so existing models keep their previous behavior.

File: megatron/core/transformer/transformer_config.py (hunk 2 above)
Comment (HollowMan6) [reply megatron/core/transformer/transformer_config.py]: Resolved. The flag is now named dsa_indexer_kpool_use_quantization and is placed directly after dsa_indexer_kpool. The old dsa_indexer_kpool_fp8 name was removed.

File: megatron/core/transformer/experimental_attention_variant/absorbed_mla.py (hunk 3 above)
Comment (HollowMan6) [reply megatron/core/transformer/experimental_attention_variant/absorbed_mla.py]: Resolved. MLA projection initialization and forward calls now share _mla_projection_context(). It returns nullcontext() unless mla_proj_disable_quantization is enabled.

File: megatron/core/transformer/experimental_attention_variant/dsa.py (hunk 4 above)
Comment (HollowMan6) [reply megatron/core/transformer/experimental_attention_variant/dsa.py:2691]: KPool currently implements pooled selection only; the existing indexer loss is token-based and has no valid pool-level target. We now reject dsa_indexer_kpool > 1 together with a positive indexer-loss coefficient during configuration validation, with a runtime guard as a defensive check. GLM-5.3 sets the loss coefficient to zero.

File: megatron/core/transformer/experimental_attention_variant/dsa.py (hunk 5 above)
Comment (HollowMan6) [reply megatron/core/transformer/experimental_attention_variant/dsa.py]: Resolved. The duplicate call was removed, and the helper is now invoked only once on the applicable path.

File: megatron/core/transformer/experimental_attention_variant/dsa.py (hunk 6 above)
Comment (HollowMan6) [reply megatron/core/transformer/experimental_attention_variant/dsa.py]: Resolved. DSA query/key projection quantization is controlled by dsa_indexer_qk_proj_disable_quantization, whose default is False, so existing models retain the enclosing quantization policy.

File: megatron/core/transformer/transformer_config.py (hunk 7 above)
Comment (HollowMan6) [reply megatron/core/transformer/transformer_config.py]: Resolved, now it's True by default.

File: megatron/core/transformer/experimental_attention_variant/dsa.py (hunk 8 above)
Comment (HollowMan6) [reply megatron/core/transformer/experimental_attention_variant/dsa.py:1648]: The fixed APE parameter now has requires_grad=False. The compression gate remains trainable intentionally: its projection is differentiable and its backward behavior is covered by the KPool projection test.

File: megatron/core/transformer/experimental_attention_variant/absorbed_mla.py (hunk 10 above)
Comment (HollowMan6) [reply megatron/core/transformer/experimental_attention_variant/absorbed_mla.py]: Resolved. The same MLA projection context wraps both parameter initialization and runtime GEMMs, so enabling mla_proj_disable_quantization preserves BF16 parameters as well as BF16 computation.

File: megatron/core/transformer/experimental_attention_variant/dsa.py (hunk 9 above)
Comment (HollowMan6) [reply megatron/core/transformer/experimental_attention_variant/dsa.py:1752]: Resolved. linear_wq_b is now invoked once through the shared projection-precision context

File: megatron/core/transformer/experimental_attention_variant/dsa.py (hunk 11 above)
Comment (HollowMan6) [reply megatron/core/transformer/experimental_attention_variant/dsa.py]: The local linear_weights_proj precision implementation was removed. After rebasing, the dsa_indexer_weights_proj_* behavior will come from #6842
